src/app/project-admin/component/project-admin.component.ts

```diff
@@ -0,0 +1,129 @@
+
+@Component({
+  selector: 'project-admin-admin',
+  templateUrl: './project-admin.component.html',
+  styleUrls: ['./project-admin.component.scss']
+})
+
+export class ProjectAdminComponent implements OnInit {
+
+  projectName = new FormControl('');
+  project$: any;
+  projectQuery$: any;
+  project: any;
+  personalAccessTokens: any;
+  personalAccessTokensQuery$: any;
+  newToken: string;
+  tokenCopied: boolean = false;
+  tokenNameIsDuplicated: boolean;
+  subscriptions$: Subscription[] = [];
+
+  constructor(
+    private routeService: RouteService,
+    private activatedRoute: ActivatedRoute,
+    private projectApolloService: ProjectApolloService,
+    private router: Router
+  ) { }
+
+  ngOnInit(): void {
+    this.checkUserIsAuthorized();
+    this.routeService.updateActivatedRoute(this.activatedRoute);
+
+    const projectId = this.activatedRoute.parent.snapshot.paramMap.get('projectId');
+    [this.projectQuery$, this.project$] = this.projectApolloService.getProjectByIdQuery(projectId);
+    this.project$.subscribe((project) => this.project = project);
+
+    NotificationService.subscribeToNotification(this, {
+      projectId: projectId,
+      whitelist: ["pat"],
+      func: this.handleWebsocketNotification
+    });
+
+    let preparationTasks$ = [];
+    preparationTasks$.push(this.preparePersonalAccessTokensRequest(projectId));
+
+  }
+
+  private checkUserIsAuthorized() {
+    if (!ConfigManager.isInit()) {
+      timer(250).subscribe(() => this.checkUserIsAuthorized());
+      return;
+    }
+
+    if (!ConfigManager.getIsAdmin()) {
+      console.log("you shouldn't be here")
+      this.router.navigate(["projects"])
+    }
+  }
+
+  preparePersonalAccessTokensRequest(projectId: string) {
+    let personalAccessTokens$;
+    [this.personalAccessTokensQuery$, personalAccessTokens$] = this.projectApolloService.getAllPersonalAccessTokens(projectId);
+    this.subscriptions$.push(personalAccessTokens$.subscribe((personalAccessTokens) => { this.personalAccessTokens = personalAccessTokens.map((token) => { return { ...token, scope: this.convertTokenScope(token.scope), expiresAt: this.convertTokenDate(token.expiresAt), createdAt: this.convertTokenDate(token.createdAt), lastUsed: this.convertTokenDate(token.lastUsed) } }) }));
+    return personalAccessTokens$;
+  }
+
+  handleWebsocketNotification(msgParts) {
+    if (msgParts[1] == 'pat') {
+      let id = msgParts[2];
+      let access_token_ids = this.personalAccessTokens.map((token) => token.id);
+      if (access_token_ids.includes(id)) {
+        this.personalAccessTokensQuery$.refetch();
+      }
+    }
+  }
+
+  deletePersonalAccessToken(tokenId: string) {
+    this.projectApolloService
+      .deletePersonalAccessTokenById(this.project.id, tokenId)
+      .pipe(first()).subscribe();
+  }
+
+  createPersonalAccessToken(name: string, expiresAt: string, scope: string) {
+    this.checkTokenNameIsDuplicated(name);
```

Comment:
i would make this a function with return value instead of writing a class variable in this function

Reply:
But i am using this class variable in the html to render a warning besides the usage in the ts file and if this would be a function with return it would get called multiple times per second wouldnt it?